src/components/ingestion/AbstractIngestionEngine.js

```diff
@@ -1,100 +1,209 @@
 import { spawn } from 'child_process'
+import {
+	pipeline,
+	Writable,
+	Transform,
+} from 'stream'
 import { Payload } from '@tvkitchen/base-classes'
 import { dataTypes } from '@tvkitchen/base-constants'
 import {
 	AbstractInstantiationError,
 	NotImplementedError,
 } from '@tvkitchen/base-errors'
+import {
+	TSDemuxer,
+} from 'ts-demuxer'
+
 import logger from '%src/lib/logger'
 import kafka from '%src/lib/kafka'
+import {
+	tsToMilliseconds,
+	generateEmptyPacket,
+} from '%src/lib/utils/mpegts'
 
+/**
+ * The AbstractIngestionEngine handles the bulk of the stream processing associated with video
+ * ingestion. Ingeston Engines that implement this class are responsible for defining the input
+ * stream, and the abstract class will handle coordinating the input pipeline that ultimately
+ * produces messages to the STREAM.CONTAINER kafka queue.
+ *
+ * The pipeline is consists of the following steps:
+ *
+ * 1. The input stream emits video data.
+ * 2. That video data is piped to an FFmpeg spawn process in order to wrap it in an MPEG-TS
+ *    container.
+ * 3. The FFmpeg spawn process emits MPEG-TS data.
+ * 4. That MPEG-TS data is processed by `TSDemuxer` in order to decorate it with the display
+ *    timestamp (DTS).
+ * 5. The decorated data is piped to Kafka as TV Kitchen `Payload`.
+ *
+ * When the stream ends, the FFMPEG process and pipeline is shut down.
```

Comment:
```suggestion
 * When the stream ends, the FFmpeg process and pipeline is shut down.
```

You did this one on purpose.

Reply:
I was probably feeling angry at the time